adds basic index release validation

xh3b4sd opened the pull request:
Towards https://github.com/giantswarm/giantswarm/issues/3047.

xh3b4sd commented on index_release.go:
```diff
@@ -0,0 +1,23 @@
+	Version     string      `yaml:"version"`
+}
+
+// TODO define and implement validation rules
```
What should be the rules? 

- Release versions should not be duplicated.
- Versions should be valid semver versions. 
- Date should not be empty. 
- Authority endpoint should not be empty.

xh3b4sd replied:
@tuommaki since I am not there tomorrow and the version bundle fixes are prio 1 for the team feel free to take over here and start implementing some validation. Doesn't have to be complete nor perfect in the end but then we can go over with the other tasks from the list and iterate towards the actual fix.

tuommaki replied:
Do we want to enforce semver format to `Version` string? Would we have use cases for `2.3.1-patchX` or possibly `2.3.1+kvm` or anything along the line?

xh3b4sd replied:
I thought so. There are several libs doing this so we should just be able to use these.

MarcelMue replied:
So only the semver validation is missing currently?

xh3b4sd replied:
And the renaming. We will come back soon.

xh3b4sd commented on index_release.go:
```diff
@@ -0,0 +1,23 @@
+
+const indexReleaseTimestampFormat = "2006-01-02T15:04:05.00Z"
+
+type IndexRelease struct {
```
I am not happy with the naming. I just chose something to get going. Please suggest alternatives.

tuommaki replied:
- `ReleaseBundle`
- `ReleaseEntry`
- `Release` - I actually like this the most. Makes members sound right: `release.Authorities` or `if release.Active {`.

xh3b4sd replied:
We already have a `Release` type which is what the `cluster-service` "aggregates".

tuommaki replied:
But when we don't create releases by aggregation anymore, could we refactor that (i.e. remove old and related stuff and use new one)?

xh3b4sd replied:
Possible that we can throw away the other types when we got rid of the aggregation magic in the end. That would be cool.

tuommaki commented on authority.go:
```diff
@@ -0,0 +1,11 @@
+type Authority struct {
+	Endpoint *url.URL `yaml:"endpoint"`
+	Name     string   `yaml:"Name"`
+	Version  string   `yaml:"version"`
```
`Provider` needs to be added here.

xh3b4sd replied:
Why do we need to add a provider?

xh3b4sd replied:
I thought there could be a map having the provider as index. I do not like to hard code the provider somehow. When you think this is right then lets move on but I feel like it is wrong.

tuommaki replied:
I would rather try to find a way to get rid of the whole field, but I don't have working idea for `cluster-operator` yet.

tuommaki commented:
Didn't touch naming but added some initial validation for releases.

xh3b4sd commented on index_release.go:
```diff
@@ -0,0 +1,117 @@
+		sha256Hash.Reset()
+		sha256Hash.Write([]byte(strings.Join(authorities, ",")))
+
+		hexHash := hex.EncodeToString(sha256Hash.Sum(nil))
```
Doesn't it work to just marshal the structure via JSON or YAML in order to get a comparable string version of the structure?

tuommaki replied:
This would break in two occasions:
 - If we change `Provider` to be a map of labels, ordering of map in Go is random and therefore would behave incorrectly here.
 - One could intentionally or accidentally form endpoint URLs that are different from string comparison point of view, but semantically or even just practically yield the same result.

I feel right now that explicitly limiting validation to only specific fields is more robust towards future changes than serialization approach.

indyfree replied:
Like the explicit approach here too :+1:

xh3b4sd commented:
> One rather important aspect we should definitely ensure is that the same release version cannot be used twice.

I just saw that this is already ensured and tested. Cool thing. I added another test case and fixed the YAML marshalling. I am fine with going with this.

xh3b4sd commented:
@giantswarm/sig-operator fishing for LGTM here. Tuomas and I worked together on this. A third pair of eyes is appreciated. Also a fourth.

indyfree commented on index_release.go:
```diff
@@ -0,0 +1,117 @@
+// TODO define and implement validation rules
+func ValidateIndexReleases(indexReleases []IndexRelease) error {
+	if len(indexReleases) == 0 {
+		return nil
```
Function looks very clean and straightforward. Just to understand the big picture: `ValidateIndexReleases` will be called by the provider-operators?

xh3b4sd replied:
We will use the validation in `opsctl` and will then use `opsctl` in the `installation` repo's CI.
